compiler/src/main/java/dagger/internal/codegen/GraphAnalysisProcessor.java

```diff
@@ -0,0 +1,388 @@
+    }
+    result.put(name, module);
+
+    // Recurse for each included module.
+    Types types = processingEnv.getTypeUtils();
+    List<Object> seedModules = new ArrayList<Object>();
+    seedModules.addAll(Arrays.asList((Object[]) annotation.get("includes")));
+    if (!annotation.get("addsTo").equals(Void.class)) seedModules.add(annotation.get("addsTo"));
+    for (Object include : seedModules) {
+      if (!(include instanceof TypeMirror)) {
+        // TODO(tbroyer): pass annotation information
+        processingEnv.getMessager().printMessage(Diagnostic.Kind.WARNING,
+            "Unexpected value for include: " + include + " in " + module, module);
+        continue;
+      }
+      TypeElement includedModule = (TypeElement) types.asElement((TypeMirror) include);
+      path.push(name);
+      collectIncludesRecursively(includedModule, result, path);
+      path.pop();
+    }
+  }
+
+  static class ProviderMethodBinding extends ProvidesBinding<Object> {
+    private final ExecutableElement method;
+    private final Binding<?>[] parameters;
+
+    protected ProviderMethodBinding(String provideKey, ExecutableElement method, boolean library) {
+      super(provideKey, method.getAnnotation(Singleton.class) != null,
+          className(method), method.getSimpleName().toString());
+      this.method = method;
+      this.parameters = new Binding[method.getParameters().size()];
+      setLibrary(library);
+    }
+
+    @Override public void attach(Linker linker) {
+      for (int i = 0; i < method.getParameters().size(); i++) {
+        VariableElement parameter = method.getParameters().get(i);
+        String parameterKey = GeneratorKeys.get(parameter);
+        parameters[i] = linker.requestBinding(parameterKey, method.toString(),
+            getClass().getClassLoader());
+      }
+    }
+
+    @Override public Object get() {
+      throw new AssertionError("Compile-time binding should never be called to inject.");
+    }
+
+    @Override public void injectMembers(Object t) {
+      throw new AssertionError("Compile-time binding should never be called to inject.");
+    }
+
+    @Override public void getDependencies(Set<Binding<?>> get, Set<Binding<?>> injectMembers) {
+      Collections.addAll(get, parameters);
+    }
+
+    @Override public String toString() {
+      return "ProvidesBinding[key=" + provideKey
+          + " method=" + moduleClass + "." + method.getSimpleName() + "()";
+    }
+  }
+
+  void writeDotFile(TypeElement module, Map<String, Binding<?>> bindings) throws IOException {
+    JavaFileManager.Location location = StandardLocation.SOURCE_OUTPUT;
+    String path = getPackage(module).getQualifiedName().toString();
+    String file = module.getQualifiedName().toString().substring(path.length() + 1) + ".dot";
+    FileObject resource = processingEnv.getFiler().createResource(location, path, file, module);
+
+    Writer writer = resource.openWriter();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *writer*. The resource is referenced in statements at the following lines: 375, 376. The resource closure statement is at line: 377. There are other execution paths that do not contain closure statements, e.g., when exception is thrown by GraphVisualizer.write. Either a) close *writer* in a try-finally block or b) close the resource by declaring *writer* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).